openvdb/openvdb/points/impl/PrincipalComponentAnalysisImpl.h

```diff
@@ -729,22 +768,100 @@ pca(PointDataGridT& points,
     {
         points::GroupFilter filter(ellipsesIdx);
         filter.reset(leafnode);
-        AttributeWriteHandle<Vec3f, NullCodec> stretchHandle(leafnode.attributeArray(strIdx));
+        AttributeWriteHandle<Vec3T, NullCodec> stretchHandle(leafnode.attributeArray(strIdx));
 
         for (Index i = 0; i < stretchHandle.size(); ++i)
         {
             if (!filter.valid(&i)) continue;
-            const Vec3f stretch = stretchHandle.get(i);
-            OPENVDB_ASSERT(stretch != Vec3f::zero());
+            const Vec3T stretch = stretchHandle.get(i);
+            OPENVDB_ASSERT(stretch != Vec3T::zero());
             const float stretchScale = 1.0f / std::cbrt(stretch.product());
             stretchHandle.set(i, stretchScale * stretch);
         }
     });
 
+    // DEPRECATED, TO REMOVE
+    if (oldRotAttr != INVALID_IDX) {
+        // Copy xform to "rotation"
+        manager.foreach([&](LeafNodeT& leafnode, size_t) {
+            // pretty inefficient way to do this, but its deprecated behaviour
+            AttributeWriteHandle<Mat3T, NullCodec> oldRotHandle(leafnode.attributeArray(oldRotAttr));
+            Mat3T* R = initPcaArrayAttribute<Mat3T>(leafnode, rotIdx, /*fill=*/false);
+            for (Index idx = 0; idx < oldRotHandle.size(); ++idx) {
+                R[idx] = oldRotHandle.get(idx);
+            }
+        });
+    }
+
+    /// 8) Covert attributes to desired format
+    /// @note  We could consolidate this method into the parallel operators above,
+    ///   but in the future we may want to do more post-processing with these
+    ///   attributes (such as monitor and manipulate volume preservation, etc)
+    ///   which will require a separate step. For now, this step is relatively
+    ///   cheap.
+    timer.start("Coverting attributes");
+    if (attrs.format == PcaAttributes::AttributeOutput::STRETCH_AND_QUATERNION)
+    {
+        manager.foreach([&](LeafNodeT& leafnode, size_t) {
+            AttributeWriteHandle<Mat3T, NullCodec> rotHandle(leafnode.attributeArray(rotIdx));
+            AttributeWriteHandle<Vec3T, NullCodec> stretchHandle(leafnode.attributeArray(strIdx));
+            PcaAttributes::QuatT* Q = initPcaArrayAttribute<PcaAttributes::QuatT>(leafnode, qutIdx, /*fill=*/false);
+            for (Index idx = 0; idx < rotHandle.size(); ++idx) {
+                const auto& rot = rotHandle.get(idx);
+                // If we're a pure reflection (partial or global, det of -1),
+                // we need to convert to a rotation for quat support. Could just
```

Comment:
If we have a negative determinant, I think we can just multiply one of the columns by {-1, -1, -1} to fix this? Because we dont actually care about the actual orientation, but instead the principal axis this defines which is the same after multiplication with -1. We dont need to update the stretch values (which should always be > 0 anyway afaik).

Reply:
Nice catch on both accounts, fixed